### ISSUE TITLE
Unison should tell the user if it infers a concrete effect in their type signature

### ISSUE THREAD

**Issue by atacratic** [unisonweb/unison#717]
(raising after discussion in #691)

```haskell
use .base

incrementP : Nat -> Nat
incrementP x = io.printLine "incrementP"
               x + 1
-- inferred type: Nat ->{io.IO} Nat               
```

This is helpful behavior - we like ability inference.  

But people might be surprised by the type inference above, if they'd thought that their signature constrained `incrementP` to be pure.  "Hey, this thing isn't as type-safe as I thought!"

Unison should emit a message when processing `incrementP`, to point out how it has interpreted the type signature.  See Paul's comments on #691.

**Comment by mitchellwrosen** [comment]
Here's a bit of an aside: in Haskell, if I was writing some monadic thing and wanted help with inferring the "abilities" I was using, I would either -

- Not write a top level type sig (not always an option)
- Just elide the constraints and have GHC complain at me

For example,

```haskell
foo :: m ()
foo = do
  () <- ask
  putStrLn "hi"
```

would complain of missing constraints `MonadIO m` and `MonadReader () m`.

In Unison I feel like I could do something similar. Writing either of these two wrong things would presumably get me the feedback I'm after:

```haskell
bar : Nat ->{} Nat
bar n =
  printLine "hi"
  n
```
```haskell
bar : Nat ->{e} Nat
bar n =
  printLine "hi"
  n
```

So is the main benefit of ability inference on a naked `->` just that I don't even have to write the `{}`/`{e}` myself? Or perhaps there are other inference implications for higher-order functions or something?

**Comment by anovstrup** [comment]
@atacratic I came across this issue after seeing #738 and after seeing [this thread](https://unisonlanguage.slack.com/archives/CLKV43YE4/p1579736230019500) on Slack.

Risking heresy, _do_ we like ability inference? Are we sure it's worth its weight? There's no way to infer other parts of the type signature, short of having the whole thing inferred... Why just abilities?

And, if we do like ability inference, maybe some of the confusion around the feature could be resolved by its syntax more explicit (and perhaps simultaneously generalizing it to other parts of the type signature?):

```
foo : Nat ->{?} Nat
foo : ? ->{} Nat
foo : Nat ->{} ?
...

**Comment by anovstrup** [comment]
Guess I need to look at #691. I'm probably commenting from ignorance.

**Comment by anovstrup** [comment]
I understand from @pchiusano's [comment on 691](https://github.com/unisonweb/unison/issues/691#issuecomment-519931233) why we'd want a type signature `A -> B` to automatically generalize to `forall e . A ->{e} B`, but the case for _implicit ability inference_ still isn't so clear to me. Maybe this explicit syntax would be an improvement (similar to above, but with automatic generalization of the "pure" syntax):
```
foo : Nat ->{?} Nat     -- infer the ability
foo : ? -> Nat              -- infer the argument type, make the function ability polymorphic
foo : Nat -> ?              -- infer the return type, make the function ability polymorphic
foo : Nat -> Nat          -- make the function ability polymorphic
foo : Nat ->{} Nat       -- nothing is inferred or generalized
```
I'm still not quite clear on the meaning or the utility of the last type signature there (`Nat ->{} Nat`).

**Comment by aryairani** [comment]
> So is the main benefit of ability inference on a naked `->` just that I don't even have to write the `{}`/`{e}` myself? Or perhaps there are other inference implications for higher-order functions or something?

Yes, a goal was to be able infer the type of `map : (a -> b) -> [a] -> [b]` to require the ability required by the input function, inferring something like: `map : (a ->{x} b) ->{} [a] ->{x} [b]` — indicating that the input function will be called by the body.

OTOH `mapGPU : (a ->{} b) -> [a] -> [b]` requires that the input function may not use abilities; presumably because the main body can’t (in this case because it’s hypothetically running on a GPU).

**Comment by aryairani** [comment]
I think we could pretty easily infer other parts of the signature too?

**Comment by anovstrup** [comment]
@aryairani I see. With the explicit syntax I proposed above, perhaps `->` (naked arrow) could mean "infer the ability type to be either polymorphic or `{}`", while `->{?}` would mean "infer the ability (without constraint)". i.e., the typer never infer concrete abilities (other than monomorphic purity) unless explicitly requested.

**Comment by anovstrup** [comment]
I wrote up an alternative proposal (#1173) that I think may be worth considering. I'd love feedback on it, if anyone's got a moment to look it over.

**Comment by aryairani** [comment]
Existing behavior: 
```
main : Int -> ()
main _ = let
  use builtin.io printLine
  printLine "Hello!"
```
![image](https://user-images.githubusercontent.com/538571/73084025-34351080-3e9a-11ea-89e2-77a2c7dc00c4.png)

Isn't Unison already telling the user it has inferred a concrete effect in their type signature?

**Comment by anovstrup** [comment]
A stronger version of this suggestion would be for Unison to **_warn_** the user if it has inferred a concrete effect in their signature (perhaps simply by coloring the inferred concrete ability type in a way that makes it clearly stand out, maybe with a "warning" color like red, orange, or yellow).

An even stronger version would be to make this a type **_error_**, which is the essence of #1173. 

A potentially more compelling example of the concern (copied from #1173) is:
```
foo : Bar -> Baz
foo bar = 
  v = someLibFn bar
  toBaz v
```

Here, there is no obvious built-in IO, but if either `someLibFn` or `toBaz` uses concrete abilities, then a concrete ability type will be inferred for `foo` as well even though the code "looks pure" (certainly to someone familiar with Haskell) and purity may well be intended.
